projects/imagenet/imagenet_experiment.py

```diff
@@ -172,51 +176,69 @@ def _create_optimizer(model, optimizer_class, optimizer_args, optimizer_groups=N
-        # Sort modules using "group_by" key
-        group_by = optimizer_groups["group_by"]
-        sorted_modules = sorted(model.modules(), key=group_by)
+    if batch_norm_weight_decay:
+        # No need to remove weight decay. Use same optimizer args for all parameters
+        model_params = model.parameters()
+    else:
+        # Group batch norm parameters
+        def group_by_batch_norm(module):
+            return isinstance(module, _BatchNorm)
 
-        # Group module parameters using "group_by" key
+        sorted_modules = sorted(model.modules(), key=group_by_batch_norm)
         grouped_parameters = {
-            str(k): list(itertools.chain.from_iterable(m.parameters(False) for m in g))
-            for k, g in itertools.groupby(sorted_modules, key=group_by)
+            k: list(itertools.chain.from_iterable(m.parameters(False) for m in g))
+            for k, g in itertools.groupby(sorted_modules, key=group_by_batch_norm)
         }
 
-        # Add custom optimizer parameters for each group
         model_params = []
-        parameters = optimizer_groups["parameters"]
-        for k, params in grouped_parameters.items():
-            group = copy.deepcopy(optimizer_args)
-            group.update(params=params)
-            group.update(**parameters[k])
-            model_params.append(group)
-    else:
-        model_params = model.parameters()
+        for is_bn, params in grouped_parameters.items():
+            # Group model_params
+            group_args = copy.deepcopy(optimizer_args)
+            group_args.update(params=params)
+
+            # Remove 'weight_decay' from batch norm parameters
+            if is_bn:
+                group_args.update(weight_decay=0.0)
+
+            model_params.append(group_args)
 
     return optimizer_class(model_params, **optimizer_args)
 
 
+def _init_batch_norm(model):
+    """
+    Initialize ResNet50 batch norm modules
+    See https://arxiv.org/pdf/1706.02677.pdf
+
+    :param model: Resnet 50 model
+    """
+    for m in model.modules():
+        if isinstance(m, torchvision.models.resnet.BasicBlock):
+            # initialized the last BatchNorm in each BasicBlock to 0
+            m.bn2.weight = nn.Parameter(torch.zeros_like(m.bn2.weight))
+        elif isinstance(m, torchvision.models.resnet.Bottleneck):
+            # initialized the last BatchNorm in each Bottleneck to 0
+            m.bn3.weight = nn.Parameter(torch.zeros_like(m.bn3.weight))
+        elif isinstance(m, (
+            nupic.research.frameworks.pytorch.models.resnets.BasicBlock,
+            nupic.research.frameworks.pytorch.models.resnets.Bottleneck
+        )):
+            # initialized the last BatchNorm in each BasicBlock to 0
+            *_, last_bn = filter(lambda x: isinstance(x, nn.BatchNorm2d),
+                                 m.regular_path)
+            last_bn.weight = nn.Parameter(torch.zeros_like(last_bn.weight))
+        elif isinstance(m, nn.Linear):
```

Comment:
This is not batch norm. It's initializing linear weights.

Reply:
That is correct, maybe the name is misleading, but in the paper describing the BN initialization they initialize the linear layers using a gaussian distribution with zero mean and 0.01 standard
deviation. They claim that helped with large minibatches.

Reply:
👍